cocotb/bfms.py

```diff
@@ -0,0 +1,206 @@
+###############################################################################
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+###############################################################################
+
+import os
+import importlib
+import re
+from cocotb.decorators import bfm_param_int_t
+
+import_info_l = []
+export_info_l = []
+
+def register_bfm_type(T, hdl):
+    global import_info_l
+    global export_info_l
+    
+    type_info = BfmTypeInfo(T, hdl, import_info_l.copy(), export_info_l.copy())
+    BfmMgr.inst().add_type_info(T, type_info)
+    import_info_l = []
+    export_info_l = []
+    
+def register_bfm_import_info(info):
+    info.id = len(import_info_l)
+    import_info_l.append(info)
+    
+def register_bfm_export_info(info):
+    info.id = len(export_info_l)
+    export_info_l.append(info)
+    
+def bfm_hdl_path(py_file, template):
+    return os.path.join(
+        os.path.dirname(os.path.abspath(py_file)),
+        template)
+
+class BfmMethodParamInfo():
+    '''
+    Information about a single BFM-method parameter
+    '''
+    
+    def __init__(self, pname, ptype):
+        self.pname = pname
+        self.ptype = ptype
+
+class BfmMethodInfo():
+    '''
+    Information about a single BFM method
+    - Method type
+    - User-specified parameter signature
+    '''
+    
+    def __init__(self, T, signature):
+        fullname = T.__qualname__
+        fi = T.__code__
+        
+        self.T = T
+        self.signature = []
+        self.type_info = []
+        self.id = -1
+
+        locals_idx = fullname.find("<locals>")
+        if locals_idx != -1:
+            fullname = fullname[locals_idx+len("<locals>."):]
+            
+        if fullname.find('.') == -1:
+            raise Exception("Attempting to register a global method as a BFM method")
+        
+        args = fi.co_varnames[1:fi.co_argcount]
+        if len(signature) != len(args):
+            raise Exception("Wrong number of parameter-type elements: expect " + str(len(args)) + " but received " + str(len(signature)))
+        
+        
+        for i in range(len(args)):
+            a = args[i]
+            t = signature[i]
+            self.signature.append(BfmMethodParamInfo(a, t))
+            try:
+                import simulator
+                if isinstance(t, bfm_param_int_t):
+                    if t.s:
+                        self.type_info.append(simulator.BFM_SI_PARAM)
+                    else:
+                        self.type_info.append(simulator.BFM_UI_PARAM)
+            except Exception:
+                # When we're not running in simulation, don't 
+                # worry about being able to access constants from simulation
+                self.type_info.append(None)
+
+
+class BfmTypeInfo():
```

Comment:
Because we still support python 2 (#767), all of your classes need to be changed to use `(object)`, which is only implied in python 3.

```suggestion
class BfmTypeInfo(object):
```

Either that, or we need to put this PR off until after #767.

Reply:
No longer relevant, 2.7 is no longer supported in master